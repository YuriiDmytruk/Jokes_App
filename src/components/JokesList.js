import React from 'react';

import { Row, Container } from 'react-bootstrap';
import { Joke } from './index';

const JokesList = (props) => (
  <Container>
    <Row>
      {props.jokes.map((joke) => (
        <Joke key={joke.id} joke={joke} />
      ))}
    </Row>
  </Container>
);

export default JokesList;
