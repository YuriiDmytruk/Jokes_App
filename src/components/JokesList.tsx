import * as React from 'react';
import { Container, Row } from 'react-bootstrap';
import { Joke } from './index';
import { Joke as JokeType } from '../types';

type JokesListProps = {
  jokes: JokeType[];
};

const JokesList = (props: JokesListProps): JSX.Element => (
  <Container>
    <Row>
      {props.jokes.map((joke) => (
        <Joke key={joke.id} joke={joke} />
      ))}
    </Row>
  </Container>
);

export default JokesList;
