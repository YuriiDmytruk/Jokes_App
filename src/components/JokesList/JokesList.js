import React from "react";
import Joke from "../Joke/Joke";
import { Row, Container } from "react-bootstrap";

export default function JokesList(props) {
  return (
    <Container>
      <Row>
        {props.jokes.map((e) => (
          <Joke joke={e} showAll={false} />
        ))}
      </Row>
    </Container>
  );
}
