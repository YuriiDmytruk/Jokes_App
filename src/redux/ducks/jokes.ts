import { Joke, JokesState, JokesActionFetch, JokesActionAdd, JokeDelete } from '../../types'
import { put, takeEvery, call } from 'redux-saga/effects';
import { SagaIterator } from 'redux-saga';

import { fetchJokes as fetchJokesFromApi } from '../../api';

export const ADD_JOKES = 'ADD_JOKES';
export const FETCH_JOKES = 'FETCH_JOKES';
export const SET_JOKES = 'SET_JOKES';
export const DELETE_JOKE = 'DELETE_JOKE'

export const JOKES = 'JOKES'

const defaultState: JokesState = {
  jokes: JSON.parse(localStorage.getItem(JOKES)!)
};

export const jokesReducer = (
  state: JokesState = defaultState,
  action: any
): JokesState => {
  switch (action.type) {
    case ADD_JOKES:
      return { ...state, jokes: state.jokes.concat(action.jokes || []) };
    case SET_JOKES:
      return { ...state, jokes: action.jokes };
    case DELETE_JOKE:
      return { ...state, jokes: state.jokes.filter(joke => joke.id !== action.id) };
    default:
      return state;
  }
};

export const deleteJoke = (id: number): JokeDelete => {
  return {type: DELETE_JOKE, id: id}
}

export const setJokes = (jokes: Joke[]): JokesActionAdd => {
  return { type: SET_JOKES, jokes: jokes };
}

export const addJokes = (jokes: Joke[]): JokesActionAdd => {
  return { type: ADD_JOKES, jokes: jokes };
}

export const fetchJokes = (
  jokesLastID: number,
  jokesAmount: number
): JokesActionFetch => {
  return {
    type: FETCH_JOKES,
    props: {
      jokesLastID: jokesLastID,
      jokesAmount: jokesAmount,
    }
  };
}

export function* fetchJokesWorker(action: JokesActionFetch): SagaIterator {
  const data: Joke[] = yield call(
    fetchJokesFromApi,
    action.props.jokesLastID,
    action.props.jokesAmount
  );
  yield put(addJokes(data));
}

export function* jokesWatcher() {
  yield takeEvery(FETCH_JOKES, fetchJokesWorker);
}