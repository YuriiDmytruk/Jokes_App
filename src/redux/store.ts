import { createStore, applyMiddleware, Store } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import createSagaMiddleware, { Task } from 'redux-saga';
import { jokesWatcher, jokesReducer } from './ducks/jokes';

import { JokesState } from '../types';

const sagaMiddleware = createSagaMiddleware();

export const store: Store<JokesState> = createStore(
  jokesReducer,
  composeWithDevTools(applyMiddleware(sagaMiddleware))
);

sagaMiddleware.run(jokesWatcher as () => IterableIterator<Task>);