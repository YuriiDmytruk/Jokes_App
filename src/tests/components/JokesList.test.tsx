import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { createStore } from 'redux';

import JokesList from '../../components/JokesList';

import { jokesReducer } from '../../redux/ducks/jokes';

test('should render JokesList component and match snapshot', () => {
  const store = createStore(jokesReducer, { jokes: [] });

  const { asFragment } = render(
    <Provider store={store}>
      <MemoryRouter>
        <JokesList jokes={[]} />
      </MemoryRouter>
    </Provider>
  );

  expect(asFragment()).toMatchSnapshot();
});
