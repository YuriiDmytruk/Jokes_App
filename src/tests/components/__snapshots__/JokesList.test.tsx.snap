// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders JokesList correctly 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="row"
    />
  </div>
</DocumentFragment>
`;
